nmmo/task/predicate_api.py

```diff
@@ -0,0 +1,312 @@
+      cache[self.name] = progress
+    return progress
+
+  def _reset(self, config: Config):
+    self._config = config
+    if not self.check(self._config):
+      raise InvalidConstraint()
+
+  def check(self, config: Config):
+    """ Checks whether the predicate is valid
+
+    A satisfiable predicate "makes sense" given a config
+    ie. Not trying to reach target off the map
+    """
+    if not GroupConstraint().check(config, self._subject):
+      return False
+    for i, (name, constraint) in enumerate(self._constraints):
+      if constraint is None:
+        continue
+      if i < len(self._args):
+        if not constraint.check(config, self._args[i]):
+          return False
+      elif not constraint.check(config, self._kwargs[name]):
+        return False
+    return True
+
+  def sample(self, config: Config, **overload):
+    """ Samples a concrete instance of a given task.
+    
+    Allows overloading of previous parameters.
+    """
+    # Sample Constraint
+    nargs = [arg.sample(config) if isinstance(arg, Constraint) else arg
+              for arg in self._args]
+    nkwargs = {k : v.sample(config) if isinstance(v, Constraint) else v
+                for k,v in self._kwargs.items()}
+    for i, (name, _) in enumerate(self._constraints):
+      if i < len(nargs):
+        if name in nkwargs:
+          raise InvalidPredicateDefinition("Constraints should match arguments.")
+        nkwargs[name] = nargs[i]
+      else:
+        break
+
+    for k, v in overload.items():
+      nkwargs[k] = v
+     # Result
+    return self.__class__(**nkwargs)
+
+  @abstractmethod
+  def _evaluate(self, gs: GameState) -> float:
+    """ A mapping from a game state to the desirability/progress of that state.
+        __call__() will cap its value to [0, 1]
+    """
+    raise NotImplementedError
+
+  def _make_name(self, class_name, args, kwargs) -> str:
+    name = [class_name] + \
+      list(map(arg_to_string, args)) + \
+      [f"{arg_to_string(key)}:{arg_to_string(arg)}" for key, arg in kwargs.items()]
+    name = "("+'_'.join(name).replace(' ', '')+")"
+    return name
+
+  def __str__(self):
+    return self.name
+
+  @property
+  def subject(self):
+    return self._subject
+
+  def create_task(self, task_cls: Task=None,
+                  assignee: Union[Iterable[int], int]=None,
+                  reward_multiplier=1.0) -> Task:
+    """ Creates a task from this predicate"""
+    if task_cls is None:
+      from nmmo.task.task_api import Task
+      task_cls = Task
+
+    if assignee is None:
+      # the new task is assigned to this predicate's subject
+      assignee = self._subject.agents
+
+    return task_cls(eval_fn=self, assignee=assignee, reward_multiplier=reward_multiplier)
+
+  def __and__(self, other):
+    return PAND(self, other)
+  def __or__(self, other):
+    return POR(self, other)
+  def __invert__(self):
+    return PNOT(self)
+  def __rshift__(self, other):
```

Comment:
IMPLY seems really weird for task predicates, do we really need it?

Reply:
Removed.